src/main/java/org/folio/circulation/domain/OverduePeriodCalculatorService.java

```diff
@@ -0,0 +1,134 @@
+      || dueDate.isAfter(systemTime)
+      || overdueFinePolicy.isUnknown()
+      || overdueFinePolicy.getCountClosed() == null;
+  }
+
+  private CompletableFuture<Result<Integer>> getOverdueMinutes(Loan loan, DateTime systemTime) {
+    DateTime dueDate = loan.getDueDate();
+    boolean countClosed = loan.getOverdueFinePolicy().getCountClosed();
+    if (countClosed) {
+      int overdueMinutes = minutesBetween(dueDate, systemTime).getMinutes();
+      return completedFuture(succeeded(overdueMinutes));
+    } else {
+      return calendarRepository
+        .fetchOpeningPeriodsBetweenDates(loan.getCheckoutServicePointId(), dueDate, systemTime, false)
+      .thenApply(r -> r.next(this::getOpeningDaysDurationMinutes));
+    }
+  }
+
+  private Result<Integer> getOpeningDaysDurationMinutes(List<OpeningPeriod> openingDays) {
+    return succeeded(
+      openingDays.stream()
+        .map(OpeningPeriod::getOpeningDay)
+        .mapToInt(day -> day.getAllDay() ? MINUTES_PER_DAY : getOpeningDayDurationMinutes(day))
+        .sum()
+    );
+  }
+
+  private int getOpeningDayDurationMinutes(OpeningDay openingDay) {
+    return openingDay.getOpeningHour()
+      .stream()
+      .mapToInt(this::getOpeningHourDurationMinutes)
+      .sum();
+  }
+
+  private int getOpeningHourDurationMinutes(OpeningHour openingHour) {
+    LocalTime startTime = openingHour.getStartTime();
+    LocalTime endTime = openingHour.getEndTime();
+    if (ObjectUtils.allNotNull(startTime, endTime) && endTime.isAfter(startTime)) {
+      return getMinutesOfDay(endTime) - getMinutesOfDay(startTime);
+    }
+    return ZERO_MINUTES;
+  }
+
+  private int getMinutesOfDay(LocalTime time) {
+    return time.getHourOfDay() * MINUTES_PER_HOUR + time.getMinuteOfHour();
+  }
+
+  private Result<Integer> adjustOverdueWithGracePeriod(Loan loan, int overdueMinutes) {
+    int result = shouldIgnoreGracePeriod(loan)
+      ? overdueMinutes
+      : Math.max(overdueMinutes - getGracePeriodMinutes(loan), ZERO_MINUTES);
+
+    return Result.succeeded(result);
+  }
+
+  private boolean shouldIgnoreGracePeriod(Loan loan) {
+    Boolean ignoreGracePeriodForRecalls = loan.getOverdueFinePolicy().getIgnoreGracePeriodForRecalls();
+    return ignoreGracePeriodForRecalls != null
+      && ignoreGracePeriodForRecalls
+      && loan.wasDueDateChangedByRecall();
+  }
+
+  private int getGracePeriodMinutes(Loan loan) {
+    LoanPolicy loanPolicy = loan.getLoanPolicy();
+    if (loanPolicy.isUnknown()) {
+      return ZERO_MINUTES;
+    }
+
+    int duration = loanPolicy.getGracePeriodDuration();
+    switch (loanPolicy.getGracePeriodInterval()) {
+    case MONTHS:
+      return duration * MINUTES_PER_MONTH;
+    case WEEKS:
+      return duration * MINUTES_PER_WEEK;
+    case DAYS:
+      return duration * MINUTES_PER_DAY;
+    case HOURS:
+      return duration * MINUTES_PER_HOUR;
+    case INCORRECT:
+      return ZERO_MINUTES;
+    case MINUTES:
+    default:
+      return duration;
+    }
```

Comment:
Could this use the `TimePeriod` class and introduce a `toMinutes` method that would hide much of this from the higher level logic of this class?

Reply:
I figured that it would be more appropriate in `Period` class. What do you think?

Reply:
That is fine. `Period` and `TimePeriod` need combining at some point. That is outside of this work.